File: backend/metrics.py
"""Counters + latency samples exposed at /metrics: hit rate, DB read/write
counts, write-reduction factor, p50/p95/p99."""
from collections import deque
from typing import Deque

# ---- counters ----
cache_hits = 0
cache_misses = 0
db_reads = 0            # prefix fallback queries that hit Postgres
db_writes = 0          # rows written via batch flush (the batched UPSERTs)
db_write_batches = 0   # number of flush operations (transactions)
searches_received = 0  # POST /search calls (before aggregation)

# ---- latency samples for /suggest (ms), bounded ring buffer ----
_LAT_CAP = 5000
_suggest_latency_ms: Deque[float] = deque(maxlen=_LAT_CAP)


def record_suggest_latency(ms: float) -> None:
    _suggest_latency_ms.append(ms)


def _percentile(samples, p: float) -> float:
    if not samples:
        return 0.0
    ordered = sorted(samples)
    k = max(0, min(len(ordered) - 1, int(round((p / 100.0) * (len(ordered) - 1)))))
    return round(ordered[k], 3)


def snapshot() -> dict:
    total = cache_hits + cache_misses
    hit_rate = round(cache_hits / total, 4) if total else 0.0
    # write reduction = searches received vs rows actually written to the DB
    write_reduction = round(searches_received / db_writes, 2) if db_writes else None
    return {
        "cache_hits": cache_hits,
        "cache_misses": cache_misses,
        "cache_hit_rate": hit_rate,
        "db_reads": db_reads,
        "db_writes": db_writes,
        "db_write_batches": db_write_batches,
        "searches_received": searches_received,
        "write_reduction_factor": write_reduction,
        "suggest_latency_ms": {
            "samples": len(_suggest_latency_ms),
            "p50": _percentile(_suggest_latency_ms, 50),
            "p95": _percentile(_suggest_latency_ms, 95),
            "p99": _percentile(_suggest_latency_ms, 99),
        },
    }

File: backend/metrics.js
// Counters + latency samples exposed at /metrics.
const m = {
  cacheHits: 0,
  cacheMisses: 0,
  dbReads: 0,            // suggestion-path DB reads (0 by design: trie serves misses)
  dbWrites: 0,           // rows written via batch flush
  dbWriteBatches: 0,     // flush transactions
  searchesReceived: 0,   // POST /search calls before aggregation
};

const LAT_CAP = 5000;
const latency = [];

function recordSuggestLatency(ms) {
  latency.push(ms);
  if (latency.length > LAT_CAP) latency.shift();
}

function percentile(p) {
  if (!latency.length) return 0;
  const sorted = [...latency].sort((a, b) => a - b);
  const k = Math.max(0, Math.min(sorted.length - 1, Math.round((p / 100) * (sorted.length - 1))));
  return Math.round(sorted[k] * 1000) / 1000;
}

function snapshot() {
  const total = m.cacheHits + m.cacheMisses;
  return {
    cache_hits: m.cacheHits,
    cache_misses: m.cacheMisses,
    cache_hit_rate: total ? Math.round((m.cacheHits / total) * 10000) / 10000 : 0,
    db_reads: m.dbReads,
    db_writes: m.dbWrites,
    db_write_batches: m.dbWriteBatches,
    searches_received: m.searchesReceived,
    write_reduction_factor: m.dbWrites ? Math.round((m.searchesReceived / m.dbWrites) * 100) / 100 : null,
    suggest_latency_ms: {
      samples: latency.length,
      p50: percentile(50),
      p95: percentile(95),
      p99: percentile(99),
    },
  };
}

module.exports = { m, recordSuggestLatency, snapshot };
